### DISCUSSION TITLE
Why top face workplane is created twice in QuickStart DEMO?

### DISCUSSION THREAD

**Discussion by pbjacob** [CadQuery/cadquery#1235] [Q&A]
I was going through the official [quickstart demo](https://cadquery.readthedocs.io/en/latest/quickstart.html). What confused me is line 11, where the top face is selected for the second time:

```python
height = 60.0
width = 80.0
thickness = 10.0
diameter = 22.0
padding = 12.0

# make the base
result = (cq.Workplane("XY")
     .box(height, width, thickness)
     .faces(">Z").workplane().hole(diameter)
     .faces(">Z").workplane()      # Here it is!
     .rect(height - padding, width - padding, forConstruction=True)
     .vertices().cboreHole(2.4, 4.4, 2.1)
     .edges("|Z").fillet(2.0)
)
```

Why the same face has to be selected again, in other words, why the workplane has to be created again? I tried to delete that line, of course ending up with an unwanted result. In order to get the reason, I broke the chain, and re-wrote the code in juyter as follows:

```python
base_box = cq.Workplane().box(height, width, thickness)
drill_hole = base_box.faces(">Z").workplane().hole(diameter)
select_top_face_again = drill_hole.faces(">Z").workplane()

rect1 = drill_hole.rect(height - padding, width - padding)
rect2 = select_top_face_again.rect(height - padding, width - padding)

print(drill_hole.plane.origin)
print(select_top_face_again.plane.origin)
print(rect1.plane.origin)
print(rect2.plane.origin)

show(rect1, rect2)
```

It can be seen that, in order to see the difference between before and after the workplane creation, I drew two rectanges, and checked the origins of all the Workplanes.

The result is suprising:
```
Vector: (0.0, 0.0, 5.0)
Vector: (0.0, 0.0, 5.0)
Vector: (0.0, 0.0, 5.0)
Vector: (0.0, 0.0, 5.0)
```

From the result we can see that, after drilling the hole, all the Workplanes are pysically the same planes, but the two rectangles are created based on different planes as shown below:

![result](https://github.com/pbjacob/cadquery/blob/master/doc/_static/quickstart/2rects.png?raw=true "two rects")

It seems that after drilling the hole, the rectangle is created based on (0,0,0).

So, can anyone tell me where it all went wrong? Thank you so much.

**Comment by lorenzncode** [comment]
rect (from docstring): _Make a rectangle for each item on the stack._

After the hole is cut, the item on the stack is a Compound with center (0, 0, 0).:

```py
    .faces(">Z").workplane().hole(diameter)
```

The following rect is then moved to z=0.

The QuickStart example resolves this by creating a new Workplane.  The objects list is then empty.   

Another way is to select the >Z face so the item on the stack is a Face with center (0, 0, 5).  

```py
height = 60.0
width = 80.0
thickness = 10.0
diameter = 22.0
padding = 12.0

# make the base
result = (cq.Workplane("XY")
    .box(height, width, thickness)
    .faces(">Z").workplane().hole(diameter)
    .faces(">Z")
    .rect(height - padding, width - padding, forConstruction=True)
    .vertices().cboreHole(2.4, 4.4, 2.1)
)
```


There is a note in the docstring regarding this behavior:
```
        Future Enhancements:
            * project points not in the workplane plane onto the workplane plane
```

> checked the origins of all the Workplanes

FYI, CQ-editor provides an object stack inspector that displays the origins of all Workplanes in addition to the items on the stack.

![image](https://user-images.githubusercontent.com/16394272/212552617-d2c7ba59-7c51-4ceb-abf7-1b8473087025.png)

**Reply by pbjacob** [reply]
Thank you very much for your kind and detailed answer!

> After the hole is cut, the item on the stack is a Compound with center (0, 0, 0).

Well, I think this is the key point for me. So, after the hole is cut, I print the center of the compound and the origin of the plane as follows:

```python
base_box = cq.Workplane().box(height, width, thickness)
drill_hole = base_box.faces(">Z").workplane().hole(diameter)

print(drill_hole.objects[0].Center())
print(drill_hole.plane.origin)
```

The results are:
```shell
Vector: (8.076637761073015e-16, 1.7958928729660294e-16, -5.928065633749353e-18)
Vector: (0.0, 0.0, 5.0)
```

Just as you said, the center of the compound is (0,0,0), while the origin of the plane is still (0,0,5). If I continuously draw something, no doubt it will based on (0,0,0), not (0,0,5). 

According to my superficial understanding of the docs, the operation shall be based on the origin of the plane, why it uses the center of the object instead in this situation? Then, what is the point of checking the origin of the workplane if the TRUE origin seems to be stored in the center of the object on the stack?

Sorry for my misunderstanding somewhere in the docs. Thank you so much again.

**Reply by lorenzncode** [reply]
`rect` calls `eachpoint` where: _"each item on the stack is converted into a point"_

> why it uses the center of the object instead in this situation?

`box`, `hole`, `rect`, and `cboreHole` all call `eachpoint` in this example.  

Let's consider the `cboreHole` call and maybe it will help explain the reason.  Just before `cboreHole`, the vertices of the rect (created "for construction" purpose) are selected.  There are then four Vertex shapes on the stack.  These four items share a common Workplane and origin.  In `eachpoint`, Shapes are converted into a point using `Center`.

The vertices are converted to the points:
```
(-24.0, -34.0, 5.0)
(24.0, -34.0, 5.0)
(24.0, 34.0, 5.0)
(-24.0, 34.0, 5.0)
```



You might try explicitly pushing a 2D point onto the stack for the rect.  This way there is no confusion on how a shape is converted into a point.

```py
height = 60.0
width = 80.0
thickness = 10.0
diameter = 22.0
padding = 12.0

# make the base
result = (cq.Workplane("XY")
    .box(height, width, thickness)
    .faces(">Z").workplane().hole(diameter)
    .pushPoints([(0, 0)])
    .rect(height - padding, width - padding, forConstruction=True)
    .vertices().cboreHole(2.4, 4.4, 2.1)
)
```

The end result is the same and it doesn't require the second workplane.


A 3D point with Z value not on the workplane plane will result in the same problem that you had when deleting the second workplane call.: 

```py
height = 60.0
width = 80.0
thickness = 10.0
diameter = 22.0
padding = 12.0

# make the base
result = (cq.Workplane("XY")
    .box(height, width, thickness)
    .faces(">Z").workplane().hole(diameter)
    .pushPoints([(0, 0, -5)])
    .rect(height - padding, width - padding, forConstruction=True)
    .vertices().cboreHole(2.4, 4.4, 2.1)
)
```

The rect is offset from the workplane plane.


> Sorry for my misunderstanding somewhere in the docs. Thank you so much again.

You're welcome, hope it helps!
